Pretty-print JSON output

tomasr8 opened the pull request:
Adds a new `FormattedJSON` formatter which pretty-prints the JSON output.

Not sure if we want to add a separate entry point or keep it behind a flag when running `flake8 --format=json`?

With the `DefaultJSON`:
```json
{"./venv/lib/python3.11/site-packages/setuptools/discovery.py": [{"code": "E501", "filename": "./venv/lib/python3.11/site-packages/setuptools/discovery.py", "line_number": 120, "column_number": 80, "text": "line too long (85 > 79 characters)", "physical_line": "    def _find_iter(cls, where: _Path, exclude: _Filter, include: _Filter) -> StrIter:\n"}]}
```
With pretty printing:
```json
{
  "./venv/lib/python3.11/site-packages/setuptools/discovery.py": [
    {
      "code": "E501",
      "filename": "./venv/lib/python3.11/site-packages/setuptools/discovery.py",
      "line_number": 120,
      "column_number": 80,
      "text": "line too long (85 > 79 characters)",
      "physical_line": "    def _find_iter(cls, where: _Path, exclude: _Filter, include: _Filter) -> StrIter:\n"
    }
  ]
}
```

asottile commented on src/flake8_json_reporter/reporters.py:
```diff
@@ -46,30 +62,72 @@ def finished(self, filename):
         self.files_reported_count += 1
         self.write_line("]")
 
-    def dictionary_from(self, violation):
```
I believe this breaks the CodeClimate one

tomasr8 replied:
In what way? It has its own `dictionary_from` (doesn't matter though, I made the new formatter inherit from the old one so the method stays where it was)

asottile commented on src/flake8_json_reporter/reporters.py:
```diff
@@ -46,30 +62,72 @@ def finished(self, filename):
+    return textwrap.indent(text, " " * indent)
+
+
+class FormattedJSON(base.BaseFormatter):
```
this can probably inherit the other one?

tomasr8 commented:
Should I add some tests?

sigmavirus24 commented:
You'll also need to add something like https://github.com/PyCQA/flake8-json/blob/main/setup.cfg#L36

tomasr8 commented:
Yep
> Not sure if we want to add a separate entry point or keep it behind a flag when running flake8 --format=json?

Do we want an extra entrypoint or reuse `--format=json` with an extra flag maybe?

sigmavirus24 commented:
> Should I add some tests?

Sure

sigmavirus24 commented:
> Yep
> > Not sure if we want to add a separate entry point or keep it behind a flag when running flake8 --format=json?
> 
> Do we want an extra entrypoint or reuse `--format=json` with an extra flag maybe?
> 
> 

I would use a new entry point

tomasr8 commented:
Hi, I added a new entrypoint `json-pretty` (if you have a better I'd be happy to use that).

I also added some tests. Since there weren't any before, I went with pytest so that I can do snapshot tests, but if you want to use a different testing framework I can change that :)

The tests are integrated with tox, so you can just do `tox -e pytest` or `tox -e pytest -- --snapshot-update` to update the snapshots. I also added the tests to the CI workflow.

asottile commented:
I'm personally not a fan of `pytest-snapshot` -- the indirection makes it more difficult to understand what the tests are doing.  I prefer to have the outputs directly inline with the tests so it's clear what's being tested

tomasr8 commented:
Fair enough, it's your call. I'll update the tests :) Though in cases like these I actually find it more readable to separate the outputs (which tend to be pretty large and IMO harder to read) from the test code.

tomasr8 commented:
Updated the tests - it should be gtg now :)

asottile commented on tests/test_formatted_json.py:
```diff
@@ -0,0 +1,298 @@
+    return inspect.cleandoc(output) + "\n"
+
+
+@pytest.fixture
+def formatter(monkeypatch):
+    """Return a ``FormattedJSON`` instance which writes to a StringIO buffer."""
+
+    def get_output(self):
+        return self.output_fd.getvalue()
+
+    monkeypatch.setattr(
+        FormattedJSON, "output", property(get_output), raising=False
+    )
+    options = Namespace(output_file=None, color=False, tee=False)
+    formatter = FormattedJSON(options)
+    formatter.output_fd = io.StringIO()
+    return formatter
```
we shouldn't be monkeypatching our own code -- you can use `capsys`

asottile commented on tests/test_formatted_json.py:
```diff
@@ -0,0 +1,298 @@
+from flake8_json_reporter.reporters import FormattedJSON
+
+
+def clean(output):
+    return inspect.cleandoc(output) + "\n"
```
we shouldn't be "cleaning" the output -- we should be asserting the values *exactly*

asottile commented on tox.ini:
```diff
@@ -1,11 +1,18 @@
 [tox]
 minversion = 2.3.1
-envlist = flake8,pre-commit
+envlist = flake8,pre-commit,pytest
```
the normal base env should be tests, not a separate `pytest` env

sigmavirus24 commented on tox.ini:
```diff
@@ -1,11 +1,18 @@
-envlist = flake8,pre-commit
+envlist = flake8,pre-commit,tests
 
 [testenv]
```
Why wouldn't we just use this testenv?

tomasr8 replied:
You mean like this? Sorry, I don't normally use tox so not sure what the conventions are..

```
[testenv]
deps = flake8
       pytest
commands =
    pytest tests {posargs}
```

sigmavirus24 replied:
Yes, instead of the tests enc, just use this one

sigmavirus24 reviewed (approved):
Question isn't a blocker

asottile commented on .github/workflows/main.yml:
```diff
@@ -17,6 +17,9 @@ jobs:
           toxenv: pre-commit
+        - os: ubuntu-latest
+          python: 3.9
+          toxenv: testenv
```
```suggestion
          toxenv: py
```

asottile commented on tox.ini:
```diff
@@ -1,10 +1,12 @@
 [tox]
 minversion = 2.3.1
-envlist = flake8,pre-commit
+envlist = flake8,pre-commit,tests
```
```suggestion
envlist = py,flake8,pre-commit
```

asottile commented on tests/test_formatted_json.py:
```diff
@@ -0,0 +1,305 @@
+        },
+    )
+    stdout, _ = capsys.readouterr()
+    assert (
+        stdout
+        == """\
+{
+  "main.py": [
+    {
+      "code": "E222",
+      "filename": "main.py",
+      "line_number": 42,
+      "column_number": 4,
+      "text": "multiple spaces after operator",
+      "physical_line": "x =  1"
+    }
+  ]
+}
+"""
+    )
```
I would write these as:
```
    expected = """\
...
...
"""
    assert out == expected
```

that'll be a little less weird for black

tomasr8 replied:
Thanks for the review! I applied your suggestions :)

asottile reviewed (commented):
a few tiny things -- and then I think we're good!

asottile reviewed (approved):
![](https://i.fluffy.cc/7WXqprWzDtCpbfprBZG1Hngnp7H7pfQm.gif)
